flow-model-generator/modules/flow-model-generator-core/src/main/java/io/ballerina/flowmodelgenerator/core/DataMapManager.java

```
@@ -939,119 +931,185 @@ private MappingPort getRefMappingPort(String id, String name, RefType type, Map<
-                                memberNames.add(member.moduleInfo.modulePrefix + ":" + memberPort.typeName);
-                            } else {
-                                memberNames.add(memberPort.typeName);
-                            }
-                        }
-                    }
-                    unionPort.typeName = String.join(PIPE, memberNames);
-                    if (unionType.dependentTypes == null) {
-                        return unionPort;
-                    }
-                    Map<String, RefType> dependentTypes = unionType.dependentTypes;
-                    for (Map.Entry<String, RefType> entry : dependentTypes.entrySet()) {
-                        String key = entry.getKey();
-                        RefType value = entry.getValue();
-                        getRefMappingPort(id + "." + key, key, value, visitedTypes, references);
-                    }
-                    return unionPort;
-                } else {
-                    return new MappingUnionPort(id, name, typeName, "union", type.key);
-                }
-            } else if (type.hashCode == null || type.hashCode.isEmpty()) {
-                return new MappingPort(id, name, type.typeName, type.typeName);
+        if (type.typeName == null) {
+            return createSimpleMappingPort(id, name, typeName, type);
+        }
+
+        return switch (type.typeName) {
+            case "record" -> handleRecordType(id, name, typeName, type, visitedTypes, references);
+            case "array" -> handleArrayType(id, name, type, visitedTypes, references);
+            case "enum" -> handleEnumType(id, name, typeName, type, visitedTypes, references);
+            case "union" -> handleUnionType(id, name, typeName, type, visitedTypes, references);
+            default -> (type.hashCode == null || type.hashCode.isEmpty())
+                    ? createSimpleMappingPort(id, name, type.typeName, type)
+                    : null;
```

Can this be `null`? If not, shall we throw an exception here?

---

Addressed